### ISSUE TITLE
Unexpected behaviour when syncing changes to global transform.

### ISSUE THREAD

**Issue by guillefix** [Frooxius/NeosPublic#872]
I thought I'd make a github issue of this. 

Here are my posts from discord https://discordapp.com/channels/402159838827905024/439421735474036738/741250822985351268 and https://discordapp.com/channels/402159838827905024/439422161657135106/741295641313149010

In summary: if Host user uses a node that changes global transform (e.g. Set Global Position, at the same time than another user changes the parent of the node (e.g. by grabbing or releasing), then upon synccing these changes, the node isn't given the correct global position inputed by the set global position node.
I think this happens because Set Global Position actually causes a change in local position, which if set by the 

Host user, overrides the change in local position that accompanies the change in parent by the other user (again, this happens typically when grabbing/releasing); however, the parent is set to the one set by the other user. The combined effect results in the unexpected behaviour.

I think I have a hacky workaround for my use case, but I feel this is not desired behaviour. A high level idea for fixing this could be to record changes to global transform as being different from changes to local transform, and then when applying changes, making sure changes to global transform occur after changes to parent, within a single tick.

**Comment by Frooxius** [comment]
Yes, there's not any global transform at all, changes to the global transform translate into a change in the local position which is what gets synced. This is how the system was designed and it's not possible to record changes to global transform as different, as that would fundamentally go against Neos' design. Any solutions need to be built around its design.

However it's a bit strange what's going on in your case. Normally grabbing and releasing object will change the local position too, which would result in conflict and one of them would end up winning. However since the parent here is separate, it doesn't get conflict resolved.

This would be the most elegant solution, making sure if the local position conflicts, the parent will conflict as well, even if it was changed just on one end. This is a bit lower level change that will take some underlying mechanisms to implement, but would solve the problem. Main question would be if it would be optional or if it would be invoked by certain changes (e.g. linking changed properties into "conflict groups").

Currently the best approach is to design your system so it only has a single user changing it at the time or so that global position setting is done on the user interacting with the object. Generally that's how most internal things are designed in Neos right now.

**Comment by guillefix** [comment]
But in my example if the Host (who wins the local position conflict) were to set the parent too, it would set it to be the Holder of the user who was grabbing and just released?, as that is the parent of the slot during the same update for which it's writing to the slot's position.

The issue in my case is that the object is constantly running, and I can't guarantee any particular user, but the host will be present, so that's why I'm using ?? with HostUser fallback. But I think I have an idea that will work which is to use the nearest user hand/head. That way, user changes will happen, but they will happen before any parent changes (before grabbing). Well.... except for laser grabbing... hmmm

**Comment by Frooxius** [comment]
If you're setting the parent to the same one that's currently set, it's not considered a change and thus won't invoke a conflict. You'd need to set the parent to something different and back to force it to change. I could add a thing to force change on any data model element if that would help, that way you could make sure the parent stays the same and will conflict if another user tries to change it.

**Comment by guillefix** [comment]
Hmm that could help. Though I'd need to use OnGrabbableReleased to signal the Host that the thing was relased (and then to change its parent from the side of the Host), and in my usecase I would have preferred not to need a component reference (as the slot is being accessed could be dynamic).

A way to get around that would be to get an impulse when a conflict has happened for a field (in this case parent slot). In fact that would allow for another way to make my system work, without the need of the forceChange you mentioned. I could detect that a conflict has happened in the local position, and revert the change manually.

But could you not have a new (hidden) field called GlobalPosition which when set, sets a flag in the slot saying "global position was changed", and next time local position is accessed, it checks if the flag is true, and if so, then computes the local position according to the changed global position and the current parent, and sets the flag to false. I think that would solve the above behaviour, as long as SetGlobalPosition changes that new field. This seems to me to not require major design changes except perhaps a new type of SyncElement (though I'm not familiar enough with C# to say much more).
I think this would solve the root of the problem rather than hack a solution around it?

**Comment by guillefix** [comment]
Talking about hacking solutions, I tried putting an Update delay on the WhileTrue impulse, on the hopes that perhaps if a user releases the ball, then that would go in conflict with an impulse that was triggered n updates ago, and maybe that would cause the user who released the ball to win the conflict. Doesn't work I guess because the tick time of the output impulse of the update relay is set to the tick of the output impulse, not the tick of the input impulse. 

Another useful thing, that may serve as a workaround for my case would be a node that returns a random present user that isn't the host, which would be more natural to obtain once we have some sort of User list node, but that will need collections^TM

**Comment by shiftyscales** [comment]
You can already find if a slot is being touched by a user's laser through use of TouchableData/TouchEventRelay components plugged in to the Touchable Events node, @guillefix.

**Comment by guillefix** [comment]
Thanks @shiftyscales 
Actually I just found a work around, by inspecting a beachball with physics someone gave me (whoo madee it? i wanna thank them)
Turns out they save the local position in a variable that is set locally by the user who grabs, and effectively they undo the host override i was complaining about above!

**Comment by shiftyscales** [comment]
Do we wish to rename this issue to something like "Mechanism to force change on any data model element" as per @Frooxius' suggestion above? Or close this issue, and create a new one for that? Or just close it outright, @guillefix?

**Comment by guillefix** [comment]
Well if Froox is happy with any of his or my suggestions to help alleviate this issue, we can keep it open. Otherwise, given that I found a good workaround, I am not too fussed
